Add whether the shard state fetch is pending to the allocation explain API

dakrone opened the pull request:
If the shard state fetch is still pending, this will now return a
message like:

```json
{
  "shard" : {
    "index" : "i",
    "index_uuid" : "de1W1374T4qgvUP4a9Ieaw",
    "id" : 0,
    "primary" : false
  },
  "assigned" : false,
  "shard_state_fetch_pending": true,
  "unassigned_info" : {
    "reason" : "INDEX_CREATED",
    "at" : "2016-04-26T16:34:53.227Z"
  },
  "allocation_delay_ms" : 0,
  "remaining_delay_ms" : 0,
  "nodes" : {
    "z-CbkiELT-SoWT91HIszLA" : {
      "node_name" : "Brain Cell",
      "node_attributes" : {
        "testattr" : "test"
      },
      "store" : {
        "shard_copy" : "NONE"
      },
      "final_decision" : "NO",
      "final_explanation" : "the shard state fetch is pending",
      "weight" : 5.0,
      "decisions" : [ ]
    }
  }
}
```

Adds the `shard_state_fetch_pending` field and uses the state to
influence the final decision and final explanation.

Relates to #17372

dakrone commented:
@ywelsch can you take a look at this?

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/TransportClusterAllocationExplainAction.java:
```diff
@@ -248,11 +252,12 @@ public static ClusterAllocationExplanation explainShard(ShardRouting shard, Rout
-        return new ClusterAllocationExplanation(shard.shardId(), shard.primary(),
-                shard.currentNodeId(), remainingDelayMillis, ui, explanations);
+        return new ClusterAllocationExplanation(shard.shardId(), shard.primary(), shard.currentNodeId(),
+                remainingDelayMillis, ui, allocation.hasPendingAsyncFetch(), explanations);
```
`allocation.hasPendingAsyncFetch()` will always return false here. The field that is used to determine this value is set by Primary/ReplicaShardAllocator. Even if this field were correctly set here, it would still be the wrong value to determine whether the shard can be allocated or not. The primary/replica shard allocator is only interested in knowing whether there are still pending fetches for the targeted shard id.

dakrone replied:
Ahh okay, I see what you're saying now. I will change this to use the GatewayAllocator on the actual shard to and check if there is still data pending. I'll ping you when I update the PR.

Thanks!

dakrone commented:
@ywelsch I changed the way this works, instead of checking that flag I added a
method to the `GatewayAllocator` that takes a shard and checks whether checks
are in-flight for it, let me know if you think this is a better way to solve
this.

ywelsch commented on core/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java:
```diff
@@ -62,6 +62,24 @@ public GatewayAllocator(Settings settings, final TransportNodesListGatewayStarte
+        if (primary) {
+            AsyncShardFetch<TransportNodesListGatewayStartedShards.NodeGatewayStartedShards> fetch = asyncFetchStarted.get(shardId);
+            if (fetch != null) {
+                return fetch.getNumberOfInFlightFetches() > 0;
```
In case `fetch == null` shouldn't `hasFetchPending` return `true`?

dakrone replied:
So, the issue with that is that some shards are completely assigned no problem without the async fetch actually being non-null, so if I assume that a null fetch means that it has one pending, they would show as needing a pending fetch when in fact they're already assigned and completely fine.

I think the fetch pending should only return true if there's actually a fetch pending, rather than if there is no fetch at all

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/TransportClusterAllocationExplainAction.java:
```diff
@@ -161,6 +166,9 @@ public static NodeExplanation calculateNodeExplanation(ShardRouting shard,
         if (node.getId().equals(assignedNodeId)) {
             finalDecision = ClusterAllocationExplanation.FinalDecision.ALREADY_ASSIGNED;
             finalExplanation = "the shard is already assigned to this node";
+        } else if (shard.unassigned() && hasPendingAsyncFetch) {
```
We should take allocatedPostIndexCreate into account here (see PrimaryShardAllocator / ReplicaShardAllocator), i.e.
```
if (shard.unassigned() && shard.allocatedPostIndexCreate(indexMetaData) && hasPendingAsyncFetch) {
```

For replica shards (`ReplicaShardAllocator`) it's even a little bit more complex. Shard fetching is only relevant if not all deciders say YES (i.e. throttling leads to shard store data not being fetched) . We should thus include the condition `nodeDecision.type() != Decision.Type.YES` for replica shards to be exact.

```
if (shard.primary() == false && shard.unassigned() && shard.allocatedPostIndexCreate(indexMetaData) && nodeDecision.type() != Decision.Type.YES) {
    finalExplanation = "the shard cannot be assigned because allocation deciders say " + nodeDecision.type.name();
    finalDecision = ClusterAllocationExplanation.FinalDecision.NO
} else if (shard.unassigned() && shard.allocatedPostIndexCreate(indexMetaData) && hasPendingAsyncFetch) {
   ... shard fetching
}
```

dakrone replied:
When I look at the `allocateUnassigned` method for both the `PrimaryShardAllocator`(0) and `ReplicaShardAllocator`(1), it checks if data is still pending and skips the allocation for either before doing any sort of decider check it looks like though?

(0): https://github.com/elastic/elasticsearch/blob/ee4e470f6000fda030ad1f6d9216efec70119227/core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java#L105-L111
(1): https://github.com/elastic/elasticsearch/blob/ee4e470f6000fda030ad1f6d9216efec70119227/core/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java#L82-L86

ywelsch replied:
for ReplicaShardAllocator, you're looking at the wrong method...
see https://github.com/elastic/elasticsearch/blob/ee4e470f6000fda030ad1f6d9216efec70119227/core/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java#L138

ywelsch replied:
btw I've also had my confusions with this `processExistingRecoveries` method before ;-) At some point you don't even look at method names anymore, you just expect the code to be in that place.. :-)

dakrone replied:
Oh! I totally missed that method checking the output of deciders, thanks for the explanation! I'll update the final decision/explanation statement

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/TransportClusterAllocationExplainAction.java:
```diff
@@ -161,6 +166,9 @@ public static NodeExplanation calculateNodeExplanation(ShardRouting shard,
             finalDecision = ClusterAllocationExplanation.FinalDecision.ALREADY_ASSIGNED;
             finalExplanation = "the shard is already assigned to this node";
+        } else if (shard.unassigned() && hasPendingAsyncFetch) {
+            finalExplanation = "the shard state fetch is pending";
```
I think this should explain why shard fetching is preventing allocation (otherwise this message will be too cryptic for the user). For primary shards it is to determine which nodes have a non-stale copy of the data. For replica shards it is to determine which nodes have a copy of the data and which copy shares the most data with the primary to speed up recovery.

ywelsch replied:
Looking at the overall code of `calculateNodeExplanation` again, I think we should also improve the message in case where the shard can be assigned but we have throttling. Can be a follow-up.

dakrone replied:
I'll try to change the message to be clearer to the user (I agree it doesn't make as much sense right now)

I agree on the throttling one, I'll open a follow-up for this

ywelsch commented on docs/reference/cluster/allocation-explain.asciidoc:
```diff
@@ -30,6 +30,7 @@ The response looks like:
     "primary" : false
   },
   "assigned" : false,                             <1>
+  "shard_state_fetch_pending": false,
```
Can you add a ` <1>` an an explanatory note below? (Something similar to what I wrote above why shard state/store fetching is relevant for allocation).

dakrone replied:
Yep good idea

ywelsch commented:
@dakrone I like the change of adding a small method to `GatewayAllocator` to determine shard fetching status. I've added some suggestions to make `calculateNodeExplanation` more in line with actual decision making by Primary/ReplicaShardAlloctor and improvements to the error messages displayed to the user. One more thing I'm wondering about is whether/how we can test this feature using integration tests (ClusterAllocationExplainIT).

dakrone commented:
@ywelsch I pushed some more commits for this

> One more thing I'm wondering about is whether/how we can test this feature using integration tests (ClusterAllocationExplainIT)

I'm not sure how to do this, I tried looking for tests for the async fetching, but they're all unit tests, I'm open to ideas if you have any

ywelsch commented:
I like `calculateNodeExplanation` better now. LGTM.
One way to test this is if we add a `NetworkPartition` rule that delays requests for a specific action (e.g. `internal:gateway/local/started_shards`). This can be done by generalizing `MockService.addUnresponsiveRule` with an additional `Predicate` parameter that checks if the request is to be delayed.

dakrone commented:
> One way to test this is if we add a NetworkPartition rule that delays requests for a specific action (e.g. internal:gateway/local/started_shards). This can be done by generalizing MockService.addUnresponsiveRule with an additional Predicate parameter that checks if the request is to be delayed.

That's a good idea, thanks Yannick, I'll work on this as a subsequent PR

clintongormley commented:
@dakrone was this merged? if not, please remove the version label

dakrone commented:
@clintongormley this was merged (with a merge commit no less!) but the Github hiccups that were going on yesterday caused this to be marked as closed instead of merged

clintongormley commented:
Closed by bfce901edf3a07c211b3892b93e24e95f1ae38ca
